### ISSUE TITLE
Request for support for inequality constraint equations

### ISSUE THREAD

**Issue by Peter230655** [csu-hmc/opty#156]
As I was told, opty can handle algebraic constraint equations.
Can it also handle alegebraic inequalities? What I mean is this:
Say, $a, b \in R, a < b$. Let $f (q_i)$ be a scalar function of the generalized coordinates.
Would $a < f(q_i) < b \longleftrightarrow$
$f(q_i) - b < 0$
$a - f(q_i) < 0$ be possible?

**Thanks for any help!**

**Comment by moorepants** [comment]
No, that is currently not supported, but opty can support anything that fits an NLP problem that IPOPT can solve. See the definition here: https://coin-or.github.io/Ipopt/

Someone would have to implement inequality algebraic constraints in opty. It only constructs equality constraints at the moment.

**Comment by Peter230655** [comment]
Clear thanks!
Should I close the issue, my question was answered!, or leave it open for possible future activity by someone?

**Comment by moorepants** [comment]
I changed the title and we can leave it as an open issue.

**Comment by moorepants** [comment]
I think this should be relatively easy to implement. We just need a flag to `Problem` that let you change:

$$
0 < g(x) < 0
$$

to

$$
lower < g(x) < upper
$$

Maybe something like:

```python
Problem(..., inequalities= ((3, -1.2, 5.0), (8, 0.0, np.inf)))
```

Where the first number in each tuple corresponds to the index in the equations of motion that holds $g(x)$ and the next two numbers are lower and upper bounds for that equation.

**Comment by Peter230655** [comment]
I guess, you meant $ 0 \leq g(x) \leq 0$

**Comment by moorepants** [comment]
>  Unfortunately, I would have no idea how to implement it. :-(

I don't believe that. :)

**Comment by Peter230655** [comment]
Easier than GitHub / PRs ? :-)

**Comment by Peter230655** [comment]
If I wanted to have a look at this, would I have to have a full understanding of Ipopt?
Where would I start to look in opty?
Thanks!

**Comment by moorepants** [comment]
> If I wanted to have a look at this, would I have to have a full understanding of Ipopt?

You want have to have a *full* understanding, but you will have to understand the relevant parts.

>  Where would I start to look in opty?

The ConstraintCollocator class will need to handle extra equations. I think that it only allows # equations equal to the # of states at the moment. And the Problem class will have to be modified to allow a user to specify which equations are inequalities and whether >= or <=. We should try carefully think about the API design for specifying this, so that it isn't clunky.

**Comment by Peter230655** [comment]
This sounds beyond my capabilities.

**Comment by Peter230655** [comment]
There is a 'hack way' to take care of this.
@moorepants Since there is a hack solution, should I close this issue?

**Comment by moorepants** [comment]
No, we shouldn't close things until it is truly fixed.

**Comment by Peter230655** [comment]
Clear! 
I just do not want to blow up things unnecessarily. This iswhy I asked.
